src/soc/ibm/power9/romstage.c

```diff
@@ -76,31 +76,32 @@ static inline bool is_proper_dimm(spd_raw_data spd, int slot)
 	return true;
 }
 
-static void mark_nonfunctional(int mcs, int mca)
+static void mark_nonfunctional(uint8_t chip, int mcs, int mca)
 {
-	mem_data.mcs[mcs].mca[mca].functional = false;
+	mem_data[chip].mcs[mcs].mca[mca].functional = false;
 
 	/* Propagate upwards */
-	if (mem_data.mcs[mcs].mca[mca ^ 1].functional == false) {
-		mem_data.mcs[mcs].functional = false;
-		if (mem_data.mcs[mcs ^ 1].functional == false)
+	if (mem_data[chip].mcs[mcs].mca[mca ^ 1].functional == false) {
+		mem_data[chip].mcs[mcs].functional = false;
+		if (mem_data[chip].mcs[mcs ^ 1].functional == false)
 			die("No functional MCS left");
```

Comment:
This is not critical error with 2 CPUs, one of them can potentially have no DIMMs installed
```suggestion
			printk(BIOS_INFO, "No functional MCS left on chip %d", chip);
```

Alternatively we can check it one level higher, but I haven't analyzed if this can break 1 CPU systems. It is also tested in `prepare_dimm_data()` and it should `die()` there anyway so ordinary print should be enough.

Reply:
Made it an info message and also updated `prepare_dimm_data()` to use loops.